### PR TITLE
[ConstEval] Switch to tablegen pass generation

### PR DESCRIPTION
**Pull request by marbre**
This switches the pass generation definition to tablegen. The cleanup
includes switching passes to follow the `create*Pass` naming convention
and moving data members to private.

### CODE REVIEW

**Review by hanhanW — APPROVED** [review]

**Review by hanhanW — CHANGES_REQUESTED** [review]
We probably need to set the default target to llvm-cpu or vmvx. If the default target is not available, we can signal a failure.

https://github.com/iree-org/iree/blob/75ad9370d759f4557a68a8a97ea25508aafc6f42/compiler/src/iree/compiler/ConstEval/JitGlobals.cpp#L654-L662

**Comment by marbre** [comment]
> We probably need to set the default target to llvm-cpu or vmvx. If the default target is not available, we can signal a failure.
> 
> https://github.com/iree-org/iree/blob/75ad9370d759f4557a68a8a97ea25508aafc6f42/compiler/src/iree/compiler/ConstEval/JitGlobals.cpp#L654-L662

Yes, sounds like this could be the issue, haven't looked closely into it yet. Thanks for the suggestion.

**Comment by marbre** [comment]
While we previously had
```c++
std::unique_ptr<OperationPass<ModuleOp>> createJitGlobalsPass() {
  return std::make_unique<JitGlobalsPass>(JitGlobalsOptions{});
}
```
the auto-generated `impl` generated with tablegen
```c++
std::unique_ptr<::mlir::Pass> createJitGlobalsPass() {
    return std::make_unique<DerivedT>();
  }
```
misses to initialize an empty `JitGlobalsPassOptions`. Since I couldn't find out if there is a suitable way to convince tablegen to generate a of `createJitGlobalsPass()` similar to what we had before, I instead introduced a delegating constructor.

File: compiler/src/iree/compiler/ConstEval/JitGlobals.cpp [hunk 1]
```diff
@@ -609,8 +611,11 @@ class ProgramBuilder {
   InitializationAnalysis initializationAnalysis;
 };
 
-struct JitGlobalsPass : public JitGlobalsBase<JitGlobalsPass> {
-  JitGlobalsPass(const JitGlobalsOptions &options)
+class JitGlobalsPass final : public impl::JitGlobalsPassBase<JitGlobalsPass> {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/ConstEval/JitGlobals.cpp:614]: Yeah, this really should be a class but not struct. Thanks for making the change. Can you also move the data members to private? (i.e., add `private:` to l.867)

https://github.com/iree-org/iree/blob/d25712c84d741baf1ba8e4ebeeb6151eb34df4d6/compiler/src/iree/compiler/ConstEval/JitGlobals.cpp#L867-L872

Comment (marbre) [reply compiler/src/iree/compiler/ConstEval/JitGlobals.cpp:614]: Good catch, sure! I moved the data members to private and adjusted the PR description.

I think we could also take a look into moving some of the functions to private but that might be something for a follow up.

**Review by hanhanW — APPROVED** [review]

**Pull request merged**